src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java

```diff
@@ -490,18 +537,20 @@ void publishAcknowledge(int publishMessageId) {
   /**
    * Sends PUBREC packet to server
    *
-   * @param publishMessageId identifier of the PUBLISH message to acknowledge
+   * @param publishMessage a PUBLISH message to acknowledge
    */
-  void publishReceived(int publishMessageId) {
+  void publishReceived(MqttPublishMessage publishMessage) {
 
     MqttFixedHeader fixedHeader =
       new MqttFixedHeader(MqttMessageType.PUBREC, false, AT_MOST_ONCE, false, 0);
 
     MqttMessageIdVariableHeader variableHeader =
-      MqttMessageIdVariableHeader.from(publishMessageId);
+      MqttMessageIdVariableHeader.from(publishMessage.messageId());
 
     io.netty.handler.codec.mqtt.MqttMessage pubrec = MqttMessageFactory.newMessage(fixedHeader, variableHeader, null);
 
+    qos2inbound.put(publishMessage.messageId(), new CurrentUnacknowledgedAndPublishMessage(pubrec, publishMessage));
```

Comment:
My first impression is that we can just make it simpler saving the `publishMessage` as for the outbound queus.